File: tests/typescript/unit/revocation/inmem.test.ts
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// In-memory revocation store unit tests.

import { describe, expect, it, vi } from 'vitest'
import { InMemoryRevocationStore } from '../../../../packages/revocation/ts/src/inmem.js'

describe('InMemoryRevocationStore', () => {
  it('marks sessions revoked until ttl expiry', () => {
    vi.useFakeTimers()
    try {
      const store = new InMemoryRevocationStore({ defaultTtlMs: 1_000 })
      store.markRevoked('sid-1')
      expect(store.isRevoked('sid-1')).toBe(true)
      vi.advanceTimersByTime(1_001)
      expect(store.isRevoked('sid-1')).toBe(false)
    } finally {
      vi.useRealTimers()
    }
  })

  it('fails closed when capacity is exhausted', () => {
    const store = new InMemoryRevocationStore({ maxEntries: 2 })
    store.markRevoked('sid-1')
    store.markRevoked('sid-2')
    expect(() => store.markRevoked('sid-3')).toThrow('capacity exceeded')
    expect(store.isRevoked('sid-1')).toBe(true)
    expect(store.isRevoked('sid-2')).toBe(true)
  })

  it('reaps expired entries to make room under capacity pressure', () => {
    vi.useFakeTimers()
    try {
      const store = new InMemoryRevocationStore({ maxEntries: 2, defaultTtlMs: 1_000 })
      store.markRevoked('sid-1')
      store.markRevoked('sid-2')
      vi.advanceTimersByTime(1_001)
      store.markRevoked('sid-3')
      expect(store.isRevoked('sid-3')).toBe(true)
      expect(store.size()).toBe(1)
    } finally {
      vi.useRealTimers()
    }
  })
})

File: tests/go/unit/revocation/inmem_test.go
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// In-memory revocation store tests for TTL behavior.

package revocation_test

import (
	"testing"
	"time"

	revocation "github.com/garudex-labs/caracal/packages/revocation/go"
)

func TestInMemoryStoreRevokesUntilTTLExpiry(t *testing.T) {
	store := revocation.NewInMemoryStore(10 * time.Millisecond)

	if err := store.MarkRevoked("sid-1", 0); err != nil {
		t.Fatalf("mark revoked: %v", err)
	}
	if !store.IsRevoked("sid-1") {
		t.Fatal("expected sid to be revoked")
	}
	time.Sleep(20 * time.Millisecond)
	if store.IsRevoked("sid-1") {
		t.Fatal("expected sid to expire")
	}
	if store.IsRevoked("sid-1") {
		t.Fatal("expected expired sid to stay evicted")
	}
}

func TestInMemoryStoreExplicitTTLOverridesDefault(t *testing.T) {
	store := revocation.NewInMemoryStore(time.Hour)

	if err := store.MarkRevoked("sid-1", time.Millisecond); err != nil {
		t.Fatalf("mark revoked: %v", err)
	}
	time.Sleep(10 * time.Millisecond)
	if store.IsRevoked("sid-1") {
		t.Fatal("expected explicit short ttl to expire")
	}
}

func TestInMemoryStoreTracksMonotonicDelegationEpochsUntilTTLExpiry(t *testing.T) {
	store := revocation.NewInMemoryStore(time.Hour)

	if got := store.CurrentDelegationEpoch("zone-1"); got != 0 {
		t.Fatalf("expected epoch 0, got %d", got)
	}
	if err := store.MarkDelegationEpoch("zone-1", 5, 0); err != nil {
		t.Fatalf("mark delegation epoch: %v", err)
	}
	if err := store.MarkDelegationEpoch("zone-1", 4, 0); err != nil {
		t.Fatalf("mark stale delegation epoch: %v", err)
	}
	if got := store.CurrentDelegationEpoch("zone-1"); got != 5 {
		t.Fatalf("expected epoch 5, got %d", got)
	}
	if err := store.MarkDelegationEpoch("zone-1", 6, time.Millisecond); err != nil {
		t.Fatalf("mark delegation epoch with ttl: %v", err)
	}
	if got := store.CurrentDelegationEpoch("zone-1"); got != 6 {
		t.Fatalf("expected epoch 6, got %d", got)
	}
	time.Sleep(10 * time.Millisecond)
	if got := store.CurrentDelegationEpoch("zone-1"); got != 0 {
		t.Fatalf("expected expired epoch to reset, got %d", got)
	}
}
